src/pages/api/open-api-docs-preview-v2.ts

```diff
@@ -0,0 +1,178 @@
+if (!fs.existsSync(TMP_DIR)) {
+	fs.mkdirSync(TMP_DIR)
+}
+// Set the maximum age for a file in the temporary directory
+const MAX_FILE_AGE_HOURS = 1
+const MAX_FILE_AGE_MS = 1000 * 60 * 60 * MAX_FILE_AGE_HOURS
+
+/**
+ * This API route is meant to be used in conjunction with the OpenAPI preview
+ * page, which allows authors to preview OpenAPI specs in a multi-page format.
+ *
+ * The preview page is _not_ deployed in production, and is meant to be
+ * accessed from: https://develop.hashicorp.services/open-api-docs-preview-v2
+ *
+ * This API route:
+ *
+ * - Handles `POST` requests with preview data. Preview data is generally some
+ *   OpenAPI spec JSON, plus additional configuration - everything needed to
+ *   build a multi-page preview of the OpenAPI documentation. We save this
+ *   preview data to a temporary file with a unique ID. We return the unique
+ *   ID, so that it can be used in subsequent requests to retrieve the data.
+ *
+ * - Handles `GET` requests with a unique file ID. These requests allow pages
+ *   to retrieve previously stored preview data, even if the end user reloads
+ *   the page or navigates to a new URL. This is particularly necessary as we're
+ *   trying to render a multi-page preview, with operations on separate URLs.
+ */
+export default async function handler(
+	req: NextApiRequest,
+	res: NextApiResponse
+) {
+	if (req.method === 'POST') {
+		return handlePost(req, res)
+	} else if (req.method === 'GET') {
+		return handleGet(req, res)
+	} else {
+		res.setHeader('Allow', ['POST', 'GET'])
+		res.status(405).json({ error: 'Method not allowed' })
+	}
+}
+
+/**
+ * Handle POST requests
+ *
+ * This function is responsible for:
+ * - Writing the POST'ed data to a temporary file
+ * - Returning the unique file ID to the client
+ */
+async function handlePost(req, res) {
+	// Parse the request body, which we expect to be preview data
+	const previewData = JSON.parse(req.body)
+
+	/**
+	 * Try writing out the static props to a new file in the `tmp` folder.
+	 *
+	 * If successful, we return a newly generated unique file ID to the client.
+	 * If unsuccessful, we return an error message.
+	 */
+	try {
+		const timestamp = Date.now()
+		const uniqueFileId = `ts${timestamp}_${randomUUID()}`
+		const newTempFile = getTempFilePath(TMP_DIR, uniqueFileId)
+		fs.writeFileSync(newTempFile, JSON.stringify(previewData, null, 2))
+		res.status(200).json({ uniqueFileId })
+	} catch (error) {
+		res.status(500).json({ error: error.toString() })
+	}
+}
+
+/**
+ * Handle GET requests, attempting to read stored props from `/tmp`.
+ *
+ * We also delete old files in the temporary directory on every request
+ * received to this endpoint, to ensure files don't hang around any longer
+ * than they need to for author previewing purposes.
+ */
+async function handleGet(req, res) {
+	// Delete old files in the temporary directory, to keep it clean.
+	deleteOldFiles(TMP_DIR, MAX_FILE_AGE_MS)
```

Comment:
I think we might need it to be blocking to avoid having the function execution cut off ([ref](https://github.com/vercel/next.js/discussions/14077#discussioncomment-25105)) (maybe possible to do "background tasks" in middleware now, but I'm not sure about API functions so would rather not risk it).

I did switch over to batching the deletion of old files with `Promise.all`, that might speed things up just slightly so we're blocking for less time 🫡